File: UnionPressOnSharp/UnionPressOnSharp/Forms/Presenters/WorkersPresenter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using UnionPressOnSharp.Forms.Models;
using UnionPressOnSharp.Forms.Views;
using UnionPressOnSharp.Forms.Presenters.Common;

namespace UnionPressOnSharp.Forms.Presenters
{
   public class WorkersPresenter
    {
        private IWorkersView workersView;
        private IWorkersRepository workersRepository;
        private BindingSource workersBindingSource;
        private IEnumerable<WorkersModel> workersList;


        public WorkersPresenter(IWorkersView workersView, IWorkersRepository workersRepository)
        {
            this.workersBindingSource = new BindingSource();
            this.workersView = workersView;
            this.workersRepository = workersRepository;

            this.workersView.SearchDataEvent += SearchWorker;
            this.workersView.AddNewDataEvent += AddNewWorkers;
            this.workersView.EditDataEvent += LoadSelectedWorkersToEdit;
            this.workersView.DeleteDataEvent += DeleteSelectedWorkers;
            this.workersView.SaveDataEvent += SaveWorkers;
            this.workersView.CancelEvent += CancelAction;
            this.workersView.SetWorkersListBindingSource(workersBindingSource);

            LoadAllWorkersList();


        }


        private void LoadAllWorkersList()
        {
            workersList = workersRepository.GetAllWorkers();
            workersBindingSource.DataSource = workersList;
        }
        private void SearchWorker(object sender, EventArgs e)
        {
            bool emptyValue = string.IsNullOrWhiteSpace(this.workersView.SearchValue);
            if (emptyValue == false)
                workersList = workersRepository.GetByValue(this.workersView.SearchValue);
            else workersList = workersRepository.GetAllWorkers();
            workersBindingSource.DataSource = workersList;
        }
        private void CancelAction(object sender, EventArgs e)
        {
            workersView.isEditData = false;
            CleanviewFields();
        }
        private void SaveWorkers(object sender, EventArgs e)
        {
            var workerModel = new WorkersModel();
   
            workerModel.Surname = workersView.WorkersSurname;
            workerModel.Name = workersView.WorkersName;
            workerModel.FatherName = workersView.WorkersFatherName;
            workerModel.Special = workersView.WorkersSpecial;
            workerModel.Phone = workersView.WorkersNumber;
            try
            {
                new Common.ModelDataValidation().Validate(workerModel);
                if (workersView.isEditData)//Edit model
                {
                    workersRepository.EditData(workerModel);
                    workersView.Message = "Сотрудник отредактирован";
                }
                else //Add new model
                {
                    workersRepository.AddData(workerModel);
                    workersView.Message = "Сотрудник добавлен";
                }
                workersView.isSuccessful = true;
                LoadAllWorkersList();
                CleanviewFields();
            }
            catch (Exception ex)
            {
                workersView.isSuccessful = false;
                workersView.Message = ex.Message;
            }
        }

        private void CleanviewFields()
        {
            workersView.WorkersSurname = "";
            workersView.WorkersName = "";
            workersView.WorkersFatherName = "";
            workersView.WorkersSpecial = "";
            workersView.WorkersNumber = "";
        }
        private void DeleteSelectedWorkers(object sender, EventArgs e)
        {
            try
            {
                var worker = (WorkersModel)workersBindingSource.Current;
                workersRepository.DeleteData(worker.Phone);
                workersView.isSuccessful = true;
                workersView.Message = "Сотрудник удален";
                LoadAllWorkersList();
            }
            catch (Exception)
            {
                workersView.isSuccessful = false;
                workersView.Message = "Ошибка, не удалось удалить сотрудника";
            }
        }
        private void LoadSelectedWorkersToEdit(object sender, EventArgs e)
        {
            var worker = (WorkersModel)workersBindingSource.Current;

            workersView.WorkersSurname = worker.Surname;
            workersView.WorkersName = worker.Name;
            workersView.WorkersFatherName = worker.FatherName;
            workersView.WorkersSpecial = worker.Special;
            workersView.WorkersNumber = worker.Phone;
            workersView.isEditData = true;
        }
        private void AddNewWorkers(object sender, EventArgs e)
        {
            workersView.isEditData = false;
        }




    }
}

File: UnionPressOnC/UnionPressOnC/Forms/Presenters/WorkersPresenter.cpp
#include <string>
#include "../Views/IWorkersView.h"
#include "../Models/IWorkersRepository.h"
using namespace UnionPressOnC;
using namespace  System;
using namespace  System::Collections::Generic;
using namespace  System::Windows::Forms;

namespace UnionPressOnC
{
   public class WorkersPresenter
    {
    private:
         IWorkersView workersView;
         IWorkersRepository workersRepository;
         BindingSource workersBindingSource;
         IEnumerable<WorkersModel> workersList;


    public:
        
        WorkersPresenter(IWorkersView workersView, IWorkersRepository workersRepository)
        {
            this->workersBindingSource = gcnew BindingSource();
            this->workersView = workersView;
            this->workersRepository = workersRepository;

            this->workersView.SearchDataEvent += SearchWorker;
            this->workersView.AddNewDataEvent += AddNewWorkers;
            this->workersView.EditDataEvent += LoadSelectedWorkersToEdit;
            this->workersView.DeleteDataEvent += DeleteSelectedWorkers;
            this->workersView.SaveDataEvent += SaveWorkers;
            this->workersView.CancelEvent += CancelAction;
            this->workersView.SetWorkersListBindingSource(workersBindingSource);

            LoadAllWorkersList();
        }


    private:
        
        void LoadAllWorkersList()
        {
            workersList = workersRepository.GetAllWorkers();
            workersBindingSource.DataSource = workersList;
        }

        void SearchWorker(Object sender, EventArgs e)
        {
            bool emptyValue = String::IsNullOrWhiteSpace(this->workersView.getSearchValue());
            if (emptyValue == false)
                workersList = workersRepository.GetByValue(this->workersView.getSearchValue());
            else workersList = workersRepository.GetAllWorkers();
            workersBindingSource.DataSource = workersList;
        }
        
        void CancelAction(Object sender, EventArgs e)
        {
            workersView.setIsEditData(false);
            CleanviewFields();
        }
        
        void SaveWorkers(Object sender, EventArgs e)
        {
            WorkersModel workerModel;

            workerModel.surname = workersView.getWorkersSurname();
            workerModel.name = workersView.getWorkersName();
            workerModel.fatherName = workersView.getWorkersFatherName();
            workerModel.special = workersView.getWorkersSpecial();
            workerModel.phoneNumber = workersView.getWorkersNumber();
            try
            {
                //new Common.ModelDataValidation().Validate(workerModel);
                if (workersView.getIsEditData())//Edit model
                {
                    workersRepository.EditData(workerModel);
                    workersView.setMessage("Сотрудник отредактирован");
                }
                else //Add new model
                {
                    workersRepository.AddData(workerModel);
                    workersView.setMessage("Сотрудник добавлен");
                }
                workersView.setIsSuccessful(true);
                LoadAllWorkersList();
                CleanviewFields();
            }
            catch (Exception^ ex)
            {
                workersView.setIsSuccessful(false);
                workersView.setMessage(ex->Message);
            }
        }

        void CleanviewFields()
        {
            workersView.setWorkersSurname("");
            workersView.setWorkersName("");
            workersView.setWorkersFatherName("");
            workersView.setWorkersSpecial("");
            workersView.setWorkersNumber("");
        }

        void DeleteSelectedWorkers(Object sender, EventArgs e)
        {
            try
            {
                WorkersModel worker = (WorkersModel)workersBindingSource.Current;
                workersRepository.DeleteData(worker.phoneNumber);
                workersView.setIsSuccessful(1); //t
                workersView.setMessage("Сотрудник удален");
                LoadAllWorkersList();
            }
            catch (Exception^)
            {
                workersView.setIsSuccessful(0);
                workersView.setMessage("Ошибка, не удалось удалить сотрудника");
            }
        }
        
        void LoadSelectedWorkersToEdit(Object sender, EventArgs e)
        {
            WorkersModel worker = (WorkersModel)workersBindingSource.Current;

            workersView.setWorkersSurname(worker.surname);
            workersView.setWorkersName(worker.name);
            workersView.setWorkersFatherName(worker.fatherName);
            workersView.setWorkersSpecial(worker.special);
            workersView.setWorkersNumber(worker.phoneNumber);
            workersView.setIsEditData(1);
        }
        
        void AddNewWorkers(Object sender, EventArgs e)
        {
            workersView.setIsEditData(0);
        }




    }
}
